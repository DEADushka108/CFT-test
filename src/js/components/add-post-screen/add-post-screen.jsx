import React, {Fragment, useCallback, useEffect, useState, useRef} from 'react';
import PropTypes from 'prop-types';
import Header from '../header/header';
import Footer from '../footer/footer';
import {validateText} from '../../utils/utils';
import {AppRoute, HttpCode, PostStatus} from '../../utils/const';
import {getPostStatus} from '../../store/posts/selectors';
import {Operation as PostOperation, ActionCreator as PostCreator} from '../../store/posts/posts';
import {redirectToRoute} from '../../store/redirect/redirect';
import {connect} from 'react-redux';

const AddPostScreen = (props) => {
  const {onSuccessSubmit, onSubmit, status, onUpdatePostStatus} = props;
  const form = useRef();
  const [title, setTitle] = useState(``);
  const [text, setText] = useState(``);
  const [isValid, setValid] = useState(false);

  const handleTitleChange = useCallback((evt) => {
    setTitle(evt.target.value);
    setValid(validateText(title) && validateText(text));
  }, [title, text]);

  const handleTextChange = useCallback((evt) => {
    setText(evt.target.value);
    setValid(validateText(title) && validateText(text));
  }, [title, text]);

  const handleFormSubmit = useCallback((evt) => {
    evt.preventDefault();
    if (validateText(text) && validateText(title)) {
      onSubmit({
        title,
        userId: 1,
        body: text,
      });
      form.current.disabled = true;
      return;
    }
    onUpdatePostStatus(PostStatus.INVALID);
  }, [status, text, title]);

  useEffect(() => {
    if (status === HttpCode.OK) {
      onUpdatePostStatus(PostStatus.VALID);
      onSuccessSubmit(`${AppRoute.ROOT}`);
      return;
    }
    return;
  }, [status]);

  return <Fragment>
    <Header/>
    <main className="page-main">
      <h1 className="visually-hidden">Add new post</h1>
      <section className="add-post">
        {status === HttpCode.SERVER_ERROR &&
        <p className="add-post__text">
          Error {status} occurred. Please try again later.
        </p>}
        {status === HttpCode.BAD_REQUEST &&
        <p className="add-post__text">
          Somthing goes wrong. Please check the Internet connection and try again later.
        </p>}
        {status === PostStatus.INVALID &&
        <p className="add-post__text">
          Please type title and write some word.
        </p>}
        <form className="add-post__form" ref={form} onSubmit={handleFormSubmit}>
          <fieldset className="add-post__item post-info">
            <legend className="visually-hidden">Write title an post</legend>
            <p className="post-info__item">
              <label className="post-info__label" htmlFor="post-title">Post title</label>
              <input className="post-info__input" id="post-title" type="text" name="post-title" placeholder="Title" required onChange={handleTitleChange}/>
            </p>
            <p className="post-info__item">
              <label className="post-info__label" htmlFor="comment-field">Text</label>
              <textarea className="post-info__input" name="comment" id="comment-field" rows="5" placeholder="Your post" onChange={handleTextChange}></textarea>
            </p>
          </fieldset>
          <button className="add-post__button" type="submit" disabled={!isValid}>Submit</button>
        </form>
      </section>
    </main>
    <Footer/>

  </Fragment>;
};

AddPostScreen.propTypes = {
  status: PropTypes.number,
  onSubmit: PropTypes.func.isRequired,
  onSuccessSubmit: PropTypes.func.isRequired,
  onUpdatePostStatus: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  status: getPostStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onSubmit(post) {
    dispatch(PostOperation.addPost(post));
  },
  onSuccessSubmit(route) {
    dispatch(redirectToRoute(route));
  },
  onUpdatePostStatus(status) {
    dispatch(PostCreator.updateStatus(status));
  },
});

export {AddPostScreen};
export default connect(mapStateToProps, mapDispatchToProps)(AddPostScreen);
