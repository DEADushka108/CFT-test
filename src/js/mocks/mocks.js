export * from './const/const';
export * from './posts/posts';
export * from './store/store';
export * from './users/users';
export * from './comments/comments';
