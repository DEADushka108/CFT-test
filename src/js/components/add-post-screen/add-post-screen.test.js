import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import {HttpCode, noop, store} from '../../mocks/mocks';
import {AddPostScreen} from './add-post-screen.jsx';

it(`Should render AddPostScreen component correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <AddPostScreen
              status={HttpCode.OK}
              onSubmit={noop}
              onSuccessSubmit={noop}
              onUpdatePostStatus={noop}
            />
          </MemoryRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
