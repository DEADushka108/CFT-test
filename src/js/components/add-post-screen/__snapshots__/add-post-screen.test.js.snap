// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddPostScreen component correctly 1`] = `
Array [
  <header
    className="page-header"
  >
    <div
      className="page-header__wrapper"
    >
      <a
        className="page-header__logo"
        href="/"
        onClick={[Function]}
      >
        Chirrup
      </a>
      <nav
        className="main-nav"
      >
        <ul
          className="main-nav__list"
        >
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link main-nav__link--active"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#home-page"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Main
              </span>
            </a>
          </li>
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link"
              href="/users"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#social-group"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Users
              </span>
            </a>
          </li>
          <li
            className="main-nav__item"
          >
            <a
              className="main-nav__link"
              href="/contacts"
              onClick={[Function]}
            >
              <svg
                className="main-nav__icon"
              >
                <use
                  xlinkHref="#message"
                />
              </svg>
              <span
                className="main-nav__text"
              >
                Contact us
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>,
  <main
    className="page-main"
  >
    <h1
      className="visually-hidden"
    >
      Add new post
    </h1>
    <section
      className="add-post"
    >
      <form
        className="add-post__form"
        onSubmit={[Function]}
      >
        <fieldset
          className="add-post__item post-info"
        >
          <legend
            className="visually-hidden"
          >
            Write title an post
          </legend>
          <p
            className="post-info__item"
          >
            <label
              className="post-info__label"
              htmlFor="post-title"
            >
              Post title
            </label>
            <input
              className="post-info__input"
              id="post-title"
              name="post-title"
              onChange={[Function]}
              placeholder="Title"
              required={true}
              type="text"
            />
          </p>
          <p
            className="post-info__item"
          >
            <label
              className="post-info__label"
              htmlFor="comment-field"
            >
              Text
            </label>
            <textarea
              className="post-info__input"
              id="comment-field"
              name="comment"
              onChange={[Function]}
              placeholder="Your post"
              rows="5"
            />
          </p>
        </fieldset>
        <button
          className="add-post__button"
          disabled={true}
          type="submit"
        >
          Submit
        </button>
      </form>
    </section>
  </main>,
  <footer
    className="page-footer"
  >
    <div
      className="page-footer__wrapper"
    >
      <a
        className="page-footer__logo"
        href="/"
        onClick={[Function]}
      >
        Chirrup
      </a>
      <div
        className="page-footer__copyright"
      >
        <p
          className="page-footer__text"
        >
          Â© 2020 Chirrup Ltd.
        </p>
      </div>
    </div>
  </footer>,
]
`;
